invoices/invoiceregistry_test.go

```diff
@@ -654,18 +542,77 @@ func TestUnknownInvoice(t *testing.T) {
 	hodlChan := make(chan interface{})
 	amt := lnwire.MilliSatoshi(100000)
 	_, err := ctx.registry.NotifyExitHopHtlc(
-		hash, amt, testHtlcExpiry, testCurrentHeight,
+		testInvoicePaymentHash, amt, testHtlcExpiry, testCurrentHeight,
 		getCircuitKey(0), hodlChan, testPayload,
 	)
 	if err != channeldb.ErrInvoiceNotFound {
 		t.Fatal("expected invoice not found error")
 	}
 }
 
-type mockPayload struct {
-	mpp *record.MPP
-}
+// Tests that invoices are canceled after expiration.
+func TestInvoiceExpiryWithRegistry(t *testing.T) {
+	t.Parallel()
 
-func (p *mockPayload) MultiPath() *record.MPP {
-	return p.mpp
+	ctx := newTestContext(t)
+	defer ctx.cleanup()
+
+	// Subscribe to the not yet existing invoice.
+	subscription, err := ctx.registry.SubscribeSingleInvoice(
+		testInvoicePaymentHash,
+	)
+
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer subscription.Cancel()
+
+	if subscription.hash != testInvoicePaymentHash {
+		t.Fatalf("expected subscription for provided hash")
+	}
+
+	testExpiry := time.Hour
+
+	// Create the invoice that has already been expired.
+	timestamp := time.Date(2017, time.July, 21, 10, 0, 0, 0, time.UTC)
+	expiringInvoice := newTestInvoice(t, timestamp, testExpiry)
+
+	// Add the invoice and expect that AddInvoice is successful.
+	addIdx, err := ctx.registry.AddInvoice(
+		expiringInvoice, testInvoicePaymentHash,
+	)
```

Comment:
I would at least test the normal flow here as well where an invoice is not yet expired.

Reply:
After some thinking I decided to test expiry but not not through the subscribing for invoice state changes as it's quiet difficult due to all invoices requiring their own subscriptions. It is also problematic as with subscriptions as there's a possibility that state change remains unobserved when subscribing after `InvoiceRegisty.Start()` (cancellation might happen before the subscription is active).

I wonder if there's any simpler but also more correct ways to test this.

Reply:
What I mean is that I miss manipulation the test clock to let invoices expire.